### ISSUE TITLE
Episodes without air date: new Watch state each time script is run

### ISSUE THREAD

**Issue by Rouzax** [Taxel/PlexTraktSync#29]
When running the script it is setting episodes to watched in Trakt each time.
This is from shows in my Kids library.
For me I don't even want to sync this to Trakt but currently there is no way to exclude libraries when running. 
Would prefer the ability to exclude a library from being synced as requested in: https://github.com/Taxel/PlexTraktSync/issues/22

When running the script it will add a new watched state to Trakt from Plex while I did not watch it in the time between running the scripts.
For these shows:

- Woezel en Pip
- Sprookjesboom

For example:
Woezel en Pip - S01E01

This is the play info from Plex:
![image](https://user-images.githubusercontent.com/4103090/80858404-6accb500-8c59-11ea-8aa4-462627a6e91f.png)

This is now my Trakt history of that same episode (I've run the script 4 times).
![image](https://user-images.githubusercontent.com/4103090/80858443-9485dc00-8c59-11ea-88c1-4a175267f0f7.png)

```
2020-05-01 16:30:38,650 DEBUG:Show [Woezel en Pip (2010)]: Started sync
2020-05-01 16:30:38,650 DEBUG:get: https://api-v2launch.trakt.tv/search/tvdb/250127
2020-05-01 16:30:38,650 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': 'APIKEY', 'Authorization': 'Bearer KEY'}
2020-05-01 16:30:38,650 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/search/tvdb/250127
2020-05-01 16:30:38,657 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/search/tvdb/250127): <Response [200]>
2020-05-01 16:30:38,657 DEBUG:get: https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/watched
2020-05-01 16:30:38,657 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': 'APIKEY', 'Authorization': 'Bearer KEY'}
2020-05-01 16:30:38,657 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/watched
2020-05-01 16:30:38,659 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-05-01 16:30:39,062 DEBUG:https://api-v2launch.trakt.tv:443 "GET /shows/woezel-en-pip/progress/watched HTTP/1.1" 200 270
2020-05-01 16:30:39,067 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/watched): <Response [200]>
2020-05-01 16:30:39,067 DEBUG:get: https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/collection
2020-05-01 16:30:39,068 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': 'APIKEY', 'Authorization': 'Bearer KEY'}
2020-05-01 16:30:39,068 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/collection
2020-05-01 16:30:39,071 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-05-01 16:30:39,370 DEBUG:https://api-v2launch.trakt.tv:443 "GET /shows/woezel-en-pip/progress/collection HTTP/1.1" 200 265
2020-05-01 16:30:39,373 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/shows/woezel-en-pip/progress/collection): <Response [200]>
2020-05-01 16:30:39,373 DEBUG:get: https://api-v2launch.trakt.tv/shows/woezel-en-pip/seasons?extended=episodes
2020-05-01 16:30:39,373 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': 'APIKEY', 'Authorization': 'Bearer KEY'}
2020-05-01 16:30:39,373 DEBUG:method, url :: get, https://api-v2launch.trakt.tv/shows/woezel-en-pip/seasons?extended=episodes
2020-05-01 16:30:39,375 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-05-01 16:30:40,169 DEBUG:https://api-v2launch.trakt.tv:443 "GET /shows/woezel-en-pip/seasons?extended=episodes HTTP/1.1" 200 1126
2020-05-01 16:30:40,173 DEBUG:RESPONSE [get] (https://api-v2launch.trakt.tv/shows/woezel-en-pip/seasons?extended=episodes): <Response [200]>
2020-05-01 16:30:40,175 DEBUG:Show [Woezel en Pip (2010)]: Generated LUT in 0.8026101589202881 seconds
2020-05-01 16:30:40,201 DEBUG:http://plex.home.lan:32400 "GET /library/metadata/118844/allLeaves HTTP/1.1" 200 3023
2020-05-01 16:30:40,220 DEBUG:post: https://api-v2launch.trakt.tv/sync/history
2020-05-01 16:30:40,220 DEBUG:headers: {'Content-Type': 'application/json', 'trakt-api-version': '2', 'trakt-api-key': 'APIKEY', 'Authorization': 'Bearer KEY'}
2020-05-01 16:30:40,220 DEBUG:method, url :: post, https://api-v2launch.trakt.tv/sync/history
2020-05-01 16:30:40,222 DEBUG:Starting new HTTPS connection (1): api-v2launch.trakt.tv:443
2020-05-01 16:30:40,559 DEBUG:https://api-v2launch.trakt.tv:443 "POST /sync/history HTTP/1.1" 201 100
2020-05-01 16:30:40,564 DEBUG:RESPONSE [post] (https://api-v2launch.trakt.tv/sync/history): <Response [201]>
2020-05-01 16:30:40,564 INFO:Show [Woezel en Pip (2010)]: Marked as watched on trakt: episode S01E01
2020-05-01 16:30:40,564 DEBUG:Show [Woezel en Pip (2010)]: Synced episode S01E01
```

The last debug log:
[last_update.txt](https://github.com/Taxel/PlexTraktSync/files/4564733/last_update.txt)

I've looked through my Plex logs but don't see anything that corresponds with the run of the script.

**Comment by twolaw** [comment]
This seems the same issue as #10 😄 
Whatever, did you try to remove the cache file `trakt_cache.sqlite` and check if a sync add another watchtime to trakt after that ?
You said with a screenshot that it added watch status to all season 4, all season 5 while you did not watched those episodes, right ?

Can you confirm you face 2 problems here : 
- s01e01 watched only one time, but watch status added at each sync
- s04 and s05 collected in plex, never watched in plex, but watch status added one time in trakt ?

You don't explain things, it's not easy to guess...

**Comment by twolaw** [comment]
I think I found the problem (to be confirmed).
While searching corresponding tvshow in trakt, it uses the tvdb id of the show (`250127` for Woezel en Pip) but trakt returns not only Woezel en Pip, it also returns a tv episode of Pete and Gladys (s01e05) which have the same id.
The search send to trakt should filter by media type but it doesn't (line 65 in main.py):
`search = trakt.sync.search_by_id(x, id_type=provider)`
should be
`search = trakt.sync.search_by_id(x, id_type=provider, media_type='show')`
And same thing for movies search...

I suggest you to try this and tell us if it works.

**Comment by Rouzax** [comment]
Sorry, in my mind I'm explaining all the details.

First question, In Plex the S01E01 has indeed only 1 watched in history as do all episodes of all seasons from this show. The same goes for the `Sprookjesboom`

Second question, no. S04 and S05 are also already collected and watched once in Plex

**Comment by twolaw** [comment]
Yes, edit this line 65 and remove the cache file `trakt_cache.sqlite`. 
THen run a sync and check if status are added again.

**Comment by Rouzax** [comment]
> 
> 
> Yes, edit this line 65 and remove the cache file `trakt_cache.sqlite`.
> THen run a sync and check if status are added again.

But the code in line 65 is only movies, right? Not TVshows.

**Comment by twolaw** [comment]
Sorry, yes line 167 you are right.
line 65 is for movies.

**Comment by Rouzax** [comment]
Running 😄

**Comment by Rouzax** [comment]
No, it did not help. Did remove the cache.
Both shows got a new watched state for all episodes.
![image](https://user-images.githubusercontent.com/4103090/80861876-a8d5d300-8c71-11ea-90fe-137b121622a6.png)

[last_update.txt](https://github.com/Taxel/PlexTraktSync/files/4567503/last_update.txt)

**Comment by twolaw** [comment]
Ok, wrong path 😄 

This time I think (again) that I found the problem :
The problematic episodes don't have any air date in trakt database, and the trakt API don't send any watched status for these.
When the script ask `https://api.trakt.tv/shows/woezel-en-pip/progress/watched` it gets only trakt status of seasons 6 and 7 which have air date correctly set (you can check on trakt.tv), it should get trakt status of all seasons.
Same problem with Sprookjesboom.
Therefore, the script think those episodes are not watched in trakt, that's why it sends 'mark as seen' over and over.

For now, I don't know how to solve this. I consider this as a trakt bug.

**Comment by Rouzax** [comment]
In that case, is there a way to exclude libraries?
All the Kids stuff I don't really care about the watched state.

**Comment by twolaw** [comment]
Yes but nobody did it yet, discussion about excluding a library is here #22

**Comment by twolaw** [comment]
Tips: if you don't want to see this kid stuff in your recent trakt history, you could patch the `mark_as_seen()` function with the `seen_date` like I suggested [here](https://github.com/Taxel/PlexTraktSync/pull/21).

**Comment by Rouzax** [comment]
> 
> 
> Tips: if you don't want to see this kid stuff in your recent trakt history, you could patch the `mark_as_seen()` function with the `seen_date` like I suggested [here](https://github.com/Taxel/PlexTraktSync/pull/21).

That is just to correct the time issue, right?

For now, I'll just throw away both shows since nobody watches them.

**Comment by twolaw** [comment]
Yes, only time correction.
This way, adding many watched status won't bother you anymore becouse i won't be in recent history.

**Comment by Rouzax** [comment]
> 
> 
> Yes, only time correction.
> This way, adding many watched status won't bother you anymore becouse i won't be in recent history.

It will still be the current time instead of 2 hours ahead, for now till there is a way to exclude libraries, I'll drop the entire Kids TV library.
Thank you for your help and patience.

**Comment by twolaw** [comment]
One last thing, can you rename issue title like this :

> Episodes without air date: new Watch state each time script is run

**Comment by twolaw** [comment]
One possible solution would be to fallback to `https://api.trakt.tv/sync/watched/shows` when `https://api.trakt.tv/shows/show-name/progress/watched` don't give required data, but i'm not enough python expert to do it.

We should wait for @Taxel here.

**Comment by Rouzax** [comment]
> 
> 
> One possible solution would be to fallback to `https://api.trakt.tv/sync/watched/shows` when `https://api.trakt.tv/shows/show-name/progress/watched` don't give required data, but i'm not enough python expert to do it.
> 
> We should wait for @Taxel here.

I'm a complete nitwit so don't look at me 😃

**Comment by twolaw** [comment]
> How can I tell if an episode I'm synching has an airdate or not? I don't see anything in `last_update.log` that mentions "airdate".

@flammable you can see it on trakt.tv website :
![image](https://user-images.githubusercontent.com/10365746/80920576-0b61c880-8d71-11ea-85d9-4ac042bc2a74.png)

Did you find an episode without air date that synced correctly ?

**Comment by flammable** [comment]
Thanks, @twolaw. I'm synching a large number of items, so it was hard to tell - but since I'm a Trakt VIP member, I was able to export my [Trakt history](https://trakt.tv/users/flammable/history) to CSV and that contained a column for airdates. I opened it in Excel, sorted it by `episode_released`, and bypassed the movies to look for episodes.

It appears this episode is synching properly, despite not having an airdate:

https://trakt.tv/shows/frisky-dingo/seasons/0/episodes/2

`2020-05-03 00:47:53,463 INFO:Show [Frisky Dingo (2006)]: Finished sync`

I don't see anything else in `last_update.log` that indicates that the special was skipped, so I assume it's being considered for synching.

That special, plus these two I had mentioned in #10, appear to be the only episodes I'm synching that don't have airdates:

https://trakt.tv/shows/mystery-science-theater-3000/seasons/0/episodes/269
https://trakt.tv/shows/archer/seasons/0/episodes/3

However, movies don't appear to have airdates at all (just releases dates), so maybe some of the logic for synching those could be duplicated for these episodes? According to Trakt, these episodes without airdates _do_ have release dates.

**Comment by twolaw** [comment]
Yes, there is a solution for sure. I have [an idea](https://github.com/Taxel/PlexTraktSync/issues/29#issuecomment-623098015), maybe someone will code it one day. I don't think I'm able to.

Are you sure Frisky Dingo Special ep 2 has only one play in your trakt account? Is it still 'watched' on plex?
I tried it on my account (mark a Frisky Dingo special as watched on trakt, then fetch status progress of the show) and I faced the same situation as others.
